integration-test/1810-alternate-viewpoints.py

```diff
@@ -0,0 +1,153 @@
+            dsl.way(4, dsl.fit_in_tile(
+                z, x, y,
+                'LINESTRING(0.1 0.5, 0.1 0.1, 0.9 0.1, 0.9 0.9, 0.5 0.9)'), {
+                    'admin_level': '2',
+                    'boundary': 'administrative',
+                    'name': 'YY',
+                    'source': 'openstreetmap.org',
+                    'mz_boundary_from_polygon': True,  # need this for hack
+                }),
+            # this is just here to turn off maritime boundaries for everything
+            # in this tile.
+            dsl.way(5, dsl.tile_box(z, x, y), {
+                'source': 'tilezen.org',
+                'maritime_boundary': True,
+                'min_zoom': 0,
+                'kind': 'maritime',
+            }),
+        )
+
+        # should get a non-disputed section of border
+        self.assert_has_feature(
+            z, x, y, 'boundaries', {
+                'kind': 'country',
+                'name:left': 'XX',
+                'name:right': 'YY',
+            })
+
+        # should get a section recognised only by XX
+        self.assert_has_feature(
+            z, x, y, 'boundaries', {
+                'kind': 'unrecognized_country',
+                'kind:xx': 'country',
+            })
+
+        # should get a section recognised _except_ by XX
+        self.assert_has_feature(
+            z, x, y, 'boundaries', {
+                'kind': 'country',
+                'kind:xx': 'unrecognized_country',
+            })
+
+    def test_claim(self):
+        # test that a claim by countries BB & CC (and recognised by DD) is
+        # only kind:country for those countries. everyone else's view is
+        # kind: unrecognized_country.
+        #
+        # TODO: recognized_by not working yet.
+        import dsl
+
+        z, x, y = (8, 0, 0)
+
+        self.generate_fixtures(
+            dsl.way(1, dsl.tile_diagonal(z, x, y), {
+                'admin_level': '2',
+                'boundary': 'claim',
+                'name': 'Extent of CC claim',
+                'claimed_by': 'CC',
+                'disputed_by': 'AA;DD',
+            }),
+            dsl.way(2, dsl.tile_diagonal(z, x, y), {
+                'admin_level': '2',
+                'boundary': 'claim',
+                'name': 'Extent of BB claim',
+                'claimed_by': 'BB',
+                'disputed_by': 'AA',
+                'recognized_by': 'DD',
+            }),
+            dsl.way(3, dsl.tile_diagonal(z, x, y), {
+                'admin_level': '3',
+                'boundary': 'administrative',
+                'name': 'Region Name',
+                'type': 'boundary',
+            }),
+            dsl.way(4, dsl.tile_diagonal(z, x, y), {
+                'dispute': 'yes',
+                'disputed_by': 'AA',
+                'name': 'BB claim',  # note: also CC claim?!
+            }),
+        )
+
+        self.assert_has_feature(
+            z, x, y, 'boundaries', {
+                'kind': 'unrecognized_country',
```

Comment:
I wonder, if like the name fan out we did in https://github.com/tilezen/vector-datasource/pull/1454, if we should also indicate `kind:aa` and `kind:dd` = `unrecognized_claim`?

Reply:
I think the situation is a little bit different for boundaries. For the names, we were removing data that had come as input, which broke multi-level fallback. Single-level fallback worked just fine.

Is a situation in which we'd need to do multi-level fallback? In this test example, would we ever do a fallback sequence `kind:aa`, `kind:bb`, `kind`?

The `disputed_by = AA;DD` on the `claim` means that `AA` and `DD` dispute the claim, not whether they do or do not agree with the locations of the borders on the ground - I've assumed any viewpoint _not_ listed in `claimed_by` or `recognized_by` doesn't agree (and will default to `kind: unrecognized_*`).

When I first prototyped this, projecting the `disputed_by` from relations onto constituent ways produced borders which were `disputed_by` pretty much everyone and disappeared from the map. So I'd be very wary of using the `disputed_by` information from `claim` relations.

Is there a real-world example of where it's useful to have a multi-level fallback projected from the `claim` relation?

Reply:
> Is a situation in which we'd need to do multi-level fallback? 

Generally countries are "aligned" with some another country in that they have a few opinions about their immediate neighbours, but lean on either ISO or the aligned country's foreign policy.

In practice someone developing in country X shouldn't need to know if Tilezen does or doesn't have data for that country's worldview (POV), but they should still indicate that as the primary POV, and ideally list out ISO or the other country Y (which we likely have in the data) as fallback, before falling back to defacto kind values.

> In this test example, would we ever do a fallback sequence kind:aa, kind:bb, kind?

I think it's more likely it'd be `kind:xx` (primary), `kind:aa` (secondary)

To be clear, I think the fallback only happens client side not server side.

> I wonder, if like the name fan out we did in #1454, if we should also indicate kind:aa and kind:dd = unrecognized_claim, since it's meaningful those countries have an opinion (versus generic default)?

Is that possible in current data structure, or would it cause problems like you described? I'm open to not doing the fan out.

Reply:
OK, cool. I added a test for this in 74b4b7fc, although the logic is getting a little bit complicated now!